### ISSUE TITLE
EXC_BAD_INSTRUCTION in cancel disk retrieve task. 

### ISSUE THREAD

**Issue by onevcat** [onevcat/Kingfisher#99]
The lib dispatch crashes on from Xcode 7 beta 6 on canceling a block. Kingfisher is using `dispatch_block_cancel` to cancel a local disk retrieving task, so calling `cancel` on a task will cause a crash in Xcode 7 beta 6.

It seems an Apple's bug with libdispatch. I tried this code in a new iOS project in beta 6, and it leads to the same issue of EXC_BAD_INSTRUCTION:

```swift
let block = dispatch_block_create(DISPATCH_BLOCK_INHERIT_QOS_CLASS) { () -> Void in
    print("Hello world")
}

dispatch_async(dispatch_get_main_queue(), block)
dispatch_block_cancel(block)
```

I dug a bit deeper into this, and found it crashes here in [queue.c](https://opensource.apple.com/source/libdispatch/libdispatch-442.1.4/src/queue.c):

```c
void
dispatch_block_cancel(dispatch_block_t db)
{
	dispatch_block_private_data_t dbpd = _dispatch_block_get_data(db);
	if (!dbpd) {
		DISPATCH_CLIENT_CRASH("Invalid block object passed to "
				"dispatch_block_cancel()");
	}
```

It seems that `_dispatch_block_get_data` is not returning a valid data chunk.

I opened a radar ([rdar://22432170](https://openradar.appspot.com/radar?id=6663758294286336)) and decided to wait for a while to see whether it could be solved soon.

**Comment by onevcat** [comment]
The problem still exists in Xcode 7 GM. So I guess I should opt-out the disk task canceling temporarily (since it is not a common scene to cancel a disk search at all).

**Comment by hankbao** [comment]
This may related to https://openradar.appspot.com/22437691

**Comment by onevcat** [comment]
@hankbao Thanks for reporting this. Yes, I believe it is the same issue.

**Comment by onevcat** [comment]
Ref commit https://github.com/onevcat/Kingfisher/commit/e271e3ed0b6461efbb6b8c4f3df040a140c4f15d

Once it gets fixed, we should add this back to Kingfisher.

**Comment by hankbao** [comment]
:+1:

**Comment by hankbao** [comment]
Any luck with this after iOS 9 officially released?

**Comment by onevcat** [comment]
@hankbao With no luck. It is still crashing.

**Comment by hankbao** [comment]
Bad Apple.

**Comment by hankbao** [comment]
Found a workaround according to an Apple engineer working on Swift compiler  https://twitter.com/jckarter/status/641963107225501696

**Comment by onevcat** [comment]
Thanks for the information.

Sure, we can get a workaround by wrapping it with Objective-C. But I am not quite willing to introduce it into Kingfisher, since it is a pure Swift implementation now. It is not a big problem with lacking of disk task cancel for this project, so I decided to opt it from Kingfisher temporarily.

**Comment by hankbao** [comment]
That's alright.

**Comment by onevcat** [comment]
This is fixed in Xcode 7.1 beta 2. So we could add it back when the new Xcode gets released.

**Comment by hankbao** [comment]
:+1:

**Comment by onevcat** [comment]
Now Xcode 7.1 was released. We could get it back and I will send a p-r on it later today.

**Comment by hankbao** [comment]
Finally ;)

**Comment by onevcat** [comment]
Fixed in #128
